swap/src/api.rs

```diff
@@ -0,0 +1,417 @@
+        server_address: Option<SocketAddr>,
+    ) -> Result<Context> {
+        let data_dir = data::data_dir_from(data, is_testnet)?;
+        let env_config = env_config_from(is_testnet);
+
+        let seed = Seed::from_file_or_generate(data_dir.as_path())
+            .context("Failed to read seed in file")?;
+
+        let bitcoin_wallet = {
+            if let Some(bitcoin) = bitcoin {
+                let (bitcoin_electrum_rpc_url, bitcoin_target_block) =
+                    bitcoin.apply_defaults(is_testnet)?;
+                Some(Arc::new(
+                    init_bitcoin_wallet(
+                        bitcoin_electrum_rpc_url,
+                        &seed,
+                        data_dir.clone(),
+                        env_config,
+                        bitcoin_target_block,
+                    )
+                    .await?,
+                ))
+            } else {
+                None
+            }
+        };
+
+        let (monero_wallet, monero_rpc_process) = {
+            if let Some(monero) = monero {
+                let monero_daemon_address = monero.apply_defaults(is_testnet);
+                let (wlt, prc) =
+                    init_monero_wallet(data_dir.clone(), monero_daemon_address, env_config).await?;
+                (Some(Arc::new(wlt)), Some(prc))
+            } else {
+                (None, None)
+            }
+        };
+
+        let tor_socks5_port = tor.map(|tor| tor.tor_socks5_port);
+
+        START.call_once(|| {
+            let _ = cli::tracing::init(debug, json, data_dir.join("logs"));
+        });
+
+        let context = Context {
+            db: open_db(data_dir.join("sqlite")).await?,
+            bitcoin_wallet,
+            monero_wallet,
+            monero_rpc_process,
+            config: Config {
+                tor_socks5_port,
+                namespace: XmrBtcNamespace::from_is_testnet(is_testnet),
+                env_config,
+                seed: Some(seed),
+                server_address,
+                debug,
+                json,
+                is_testnet,
+                data_dir,
+            },
+            swap_lock: Arc::new(SwapLock::new()),
+        };
+
+        Ok(context)
+    }
+
+    pub async fn for_harness(
+        seed: Seed,
+        env_config: EnvConfig,
+        db_path: PathBuf,
+        bob_bitcoin_wallet: Arc<bitcoin::Wallet>,
+        bob_monero_wallet: Arc<monero::Wallet>,
+    ) -> Self {
+        let config = Config::for_harness(seed, env_config);
+
+        Self {
+            bitcoin_wallet: Some(bob_bitcoin_wallet),
+            monero_wallet: Some(bob_monero_wallet),
+            config,
+            db: open_db(db_path).await.unwrap(),
+            monero_rpc_process: None,
+            swap_lock: Arc::new(SwapLock::new()),
+        }
+    }
+}
+
+impl fmt::Debug for Context {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+async fn init_bitcoin_wallet(
```

Comment:
not sure this is the best place for these methods

Reply:
Agreed. Where'd you put it?

Reply:
Also this code duplicated in in `src/bin/asb.rs`

Reply:
i think it should go in a new init module where it can be called from both places